pkg/stanza/fileconsumer/internal/reader/reader.go

```diff
@@ -69,30 +70,31 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 
 	switch r.compression {
 	case "gzip":
-		// We need to create a gzip reader each time ReadToEnd is called because the underlying
-		// SectionReader can only read a fixed window (from previous offset to EOF).
-		info, err := r.file.Stat()
+		currentEOF, err := r.createGzipReader()
 		if err != nil {
-			r.set.Logger.Error("failed to stat", zap.Error(err))
 			return
 		}
-		currentEOF := info.Size()
-
-		// use a gzip Reader with an underlying SectionReader to pick up at the last
-		// offset of a gzip compressed file
-		gzipReader, err := gzip.NewReader(io.NewSectionReader(r.file, r.Offset, currentEOF))
-		if err != nil {
-			if !errors.Is(err, io.EOF) {
-				r.set.Logger.Error("failed to create gzip reader", zap.Error(err))
-			}
-			return
-		}
-		r.reader = gzipReader
 		// Offset tracking in an uncompressed file is based on the length of emitted tokens, but in this case
 		// we need to set the offset to the end of the file.
 		defer func() {
 			r.Offset = currentEOF
 		}()
+	case "auto":
+		// TODO: Find a better solution
+		// Identifying a filename by its extension may not always be correct. We could have a compressed file without the .gz extension
+		if filepath.Ext(r.fileName) == ".gz" {
```

Comment:
I think we should consider storing the "file type" in the `reader.Metadata` so we have a faster way to evaluate this in the hot path.